File: frontend/src/api/quality.ts
import { http } from "./http";

export interface QualityCapaAction {
  id: string;
  action_type: "corrective" | "preventive";
  description: string;
  owner_id: string;
  due_date: string;
  status: "open" | "completed";
  completion_evidence?: string | null;
  completed_at?: string | null;
  completed_by?: string | null;
  overdue: boolean;
}

export interface QualityNonconformance {
  id: string;
  inspection_id?: string | null;
  supplier_quality_id?: string | null;
  supplier_id?: string | null;
  supplier_period?: string | null;
  inspection_type?: string | null;
  source_type?: string | null;
  source_id?: string | null;
  source_document_name?: string | null;
  description: string;
  status: "open" | "investigating" | "closed";
  severity: "minor" | "major" | "critical";
  disposition?: "rework" | "accept" | "scrap" | "return_to_supplier" | null;
  owner_id?: string | null;
  due_date?: string | null;
  root_cause?: string | null;
  closure_evidence?: string | null;
  closed_at?: string | null;
  closed_by?: string | null;
  overdue: boolean;
  actions: QualityCapaAction[];
}

export interface QualityInvestigationPayload {
  severity: QualityNonconformance["severity"];
  disposition: NonNullable<QualityNonconformance["disposition"]>;
  owner_id: string;
  due_date: string;
  root_cause: string;
}

export interface QualityCapaCreatePayload {
  action_type: QualityCapaAction["action_type"];
  description: string;
  owner_id: string;
  due_date: string;
}

export const createInspection = (payload: unknown) => http.post("/quality/inspections", payload);
export const listInspections = () => http.get("/quality/inspections");
export const submitInspection = (id: string, results: unknown) => http.post(`/quality/inspections/${id}/submit`, results);
export const closeInspection = (id: string, disposition: string) => http.post(`/quality/inspections/${id}/close`, { disposition });
export const listNonconformances = () => http.get<{
  code: number;
  msg: string;
  data: QualityNonconformance[];
}>("/quality/nonconformances");
export const updateNonconformanceInvestigation = (id: string, payload: QualityInvestigationPayload) =>
  http.put(`/quality/nonconformances/${id}/investigation`, payload);
export const createCapaAction = (id: string, payload: QualityCapaCreatePayload) =>
  http.post(`/quality/nonconformances/${id}/actions`, payload);
export const completeCapaAction = (id: string, completionEvidence: string) =>
  http.post(`/quality/capa-actions/${id}/complete`, { completion_evidence: completionEvidence });
export const closeNonconformance = (id: string, closureEvidence: string) =>
  http.post(`/quality/nonconformances/${id}/close`, { closure_evidence: closureEvidence });
export const listQualityPlans = () => http.get("/quality/plans");
export const createQualityPlan = (payload: unknown) => http.post("/quality/plans", payload);
export const listDefects = () => http.get("/quality/defects");
export const createDefect = (payload: unknown) => http.post("/quality/defects", payload);
export const createInspectionFromPlan = (payload: unknown) => http.post("/quality/inspections/from-plan", payload);

File: backend/app/schemas/quality.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field

class InspectionCreate(BaseModel):
    inspection_type: str = Field(min_length=1, max_length=32)
    source_type: str = Field(min_length=1, max_length=64)
    source_id: str = Field(min_length=1, max_length=36)

class QaPlanCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    items: list[dict] = Field(min_length=1)


class QaDefectCreate(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    severity: Literal["minor", "major", "critical"] = "major"
    status: Literal["active", "inactive"] = "active"


class InspectionFromPlanCreate(BaseModel):
    inspection_type: str = Field(min_length=1, max_length=32)
    source_type: str = Field(min_length=1, max_length=64)
    source_id: str = Field(min_length=1, max_length=36)
    plan_id: str = Field(min_length=1, max_length=36)
    sample_size: int = Field(gt=0, le=100000)


class InspectionResult(BaseModel):
    item: str = Field(min_length=1, max_length=128)
    value: str = Field(min_length=1, max_length=500)
    passed: bool | None = None


class InspectionClose(BaseModel):
    disposition: Literal["rework", "accept", "scrap", "return_to_supplier"]


class NonconformanceInvestigationUpdate(BaseModel):
    severity: Literal["minor", "major", "critical"]
    disposition: Literal["rework", "accept", "scrap", "return_to_supplier"]
    owner_id: str = Field(min_length=1, max_length=36)
    due_date: date
    root_cause: str = Field(min_length=2, max_length=1000)


class CapaActionCreate(BaseModel):
    action_type: Literal["corrective", "preventive"]
    description: str = Field(min_length=2, max_length=500)
    owner_id: str = Field(min_length=1, max_length=36)
    due_date: date


class CapaActionComplete(BaseModel):
    completion_evidence: str = Field(min_length=2, max_length=1000)


class NonconformanceClose(BaseModel):
    closure_evidence: str = Field(min_length=2, max_length=1000)
